src/kaleidoscope/plugin/LED-ActiveModColor.cpp

```diff
@@ -32,57 +32,67 @@ cRGB ActiveModColorEffect::highlight_color = (cRGB) {
+  uint8_t key_state) {
+  // If `key_addr` is not a physical key address, ignore it:
+  if (! key_addr.isValid()) {
     return EventHandlerResult::OK;
+  }
 
-  mod_key_count_ = 0;
-
-  for (auto key_addr : KeyAddr::all()) {
-    Key k = Layer.lookupOnActiveLayer(key_addr);
-
-    if (::OneShot.isOneShotKey(k) ||
-        (highlight_normal_modifiers_ && (
-           (k >= Key_LeftControl && k <= Key_RightGui) ||
-           (k.getFlags() == (SYNTHETIC | SWITCH_TO_KEYMAP))))) {
-      mod_keys_[mod_key_count_++] = key_addr;
+  if (keyToggledOn(key_state)) {
+    // If a key toggles on, we check its value. If it's a OneShot key,
+    // it will get highlighted. Conditionally (if
+    // `highlight_normal_modifiers_` is set), we also highlight
+    // modifier and layer-shift keys.
+    if (::OneShot.isOneShotKey(key) ||
+        (highlight_normal_modifiers_ &&
+         ((key >= Key_LeftControl && key <= Key_RightGui) ||
+          (key.getFlags() == (SYNTHETIC | SWITCH_TO_KEYMAP))))) {
+      mod_key_bits_.set(key_addr);
+    }
+  } else if (keyToggledOff(key_state)) {
+    // When a non-OneShot key toggles off, if it was being
+    // highlighted, we clear it immediately. We don't do this for
+    // OneShot keys, because they remain active after toggling off.
+    if (mod_key_bits_.read(key_addr) &&
+        (! ::OneShot.isOneShotKey(key))) {
+      mod_key_bits_.clear(key_addr);
+      ::LEDControl.refreshAt(key_addr);
     }
   }
 
   return EventHandlerResult::OK;
 }
 
 EventHandlerResult ActiveModColorEffect::beforeReportingState() {
-  if (mod_key_count_ == 0) {
-    onLayerChange();
-  }
-
-  for (uint8_t i = 0; i < mod_key_count_; i++) {
-    const KeyAddr &key_addr = mod_keys_[i];
 
-    Key k = Layer.lookupOnActiveLayer(key_addr);
+  // This loop iterates through only the `key_addr`s that have their
+  // bits in the `mod_key_bits_` bitfield set.
+  for (KeyAddr key_addr : mod_key_bits_) {
+    Key key = Layer.lookup(key_addr);
 
-    if (::OneShot.isOneShotKey(k)) {
-      if (::OneShot.isSticky(k))
+    if (::OneShot.isOneShotKey(key)) {
+      // First, we deal with OneShot keys:
+      if (::OneShot.isSticky(key)) {
+        // If the OneShot key is sticky, it gets a different color:
         ::LEDControl.setCrgbAt(key_addr, sticky_color);
-      else if (::OneShot.isActive(k))
-        ::LEDControl.setCrgbAt(key_addr, highlight_color);
-      else
-        ::LEDControl.refreshAt(key_addr);
-    } else if (k >= Key_LeftControl && k <= Key_RightGui) {
-      if (kaleidoscope::Runtime.hid().keyboard().isModifierKeyActive(k))
-        ::LEDControl.setCrgbAt(key_addr, highlight_color);
-      else
-        ::LEDControl.refreshAt(key_addr);
-    } else if (k.getFlags() == (SYNTHETIC | SWITCH_TO_KEYMAP)) {
-      uint8_t layer = k.getKeyCode();
-      if (layer >= LAYER_SHIFT_OFFSET)
-        layer -= LAYER_SHIFT_OFFSET;
-
-      if (Layer.isActive(layer))
+      } else if (::OneShot.isActive(key) || ::OneShot.isPressed(key)) {
```

Comment:
In its current form, #893 has the `active` flag unset for OneShot keys that are "pending cancellation", but the `pressed` flag is still set. This change will keep held OneShot modifiers highlighted until they're released, but due to #894, it won't work correctly without merging #893 first.

Note that the change to OneShot will allow for three different colours to be used meaningfully in a future update:
- OneShot keys that will stay active until another key is pressed (active)
- OneShot keys that will stay active until manually cancelled (sticky)
- Modifiers (including OneShots) that are being held, and will stay active until released